Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -477,15 +477,23 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
     private func decodeHTTP(ctx: ChannelHandlerContext) throws {
         // We need to refetch the cumulationBuffer on each loop as it may has changed due re-entrance calls of channelRead(...)
         while let bufferSlice = self.cumulationBuffer, bufferSlice.readableBytes > 0 {
+            // we need to get `readerIndex` and `readableBytes` now because `withVeryUnsafeBytes` owns the
+            // `ByteBuffer` exlusively.
+            let readerIndex = bufferSlice.readerIndex
+            let readableBytes = bufferSlice.readableBytes
+
             // Using withVeryUnsafeBytes here as this simplifies the calculation of the readerIndex which is relative to the baseAddress.
             let result = bufferSlice.withVeryUnsafeBytes { (pointer) -> size_t in
-                self.state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+                self.state.baseAddress = pointer.baseAddress!
                 defer {
                     self.state.baseAddress = nil
                 }
-                return self.state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count) { p in
-                    c_nio_http_parser_execute(&self.parser, &self.settings, p.advanced(by: bufferSlice.readerIndex), bufferSlice.readableBytes)
-                }
+                let httpParserPointer = UnsafeRawBufferPointer(start: pointer.baseAddress!.advanced(by: readerIndex),
+                                                               count: readableBytes).bindMemory(to: Int8.self)
```

Comment:
I don't think this is safe.

By default the ByteBuffer address is bound to `UInt8` (we do it in `allocateAndPrepareRawMemory`). This bind, according to SE-107, *rebinds* the ByteBuffer memory to `Int8`, invalidating the assumption that this is bound to `UInt8`. This is fine, but if we think we're going to bind memory to `UInt8` in the initialiser we should just use `withMemoryRebound(to:)` here to put the old binding back afterwards.